### ISSUE TITLE
Error renaming file (bug in the Crystal compiler)

### ISSUE THREAD

**Issue by gaetgu** [crystal-lang/crystal#9895]
# Bug in the Crystal compiler

I got this error while first getting the language up and running:
```console
Error renaming file: '/Users/myusername/.cache/crystal/usr-local-Cellar-crystal-0.35.1_1-src-ecr-process.cr/_main.o.tmp' -> '/Users/myusername/.cache/crystal/usr-local-Cellar-crystal-0.35.1_1-src-ecr-process.cr/_main.o': No such file or directory (File::NotFoundError)
```

At the bottom it says:
```console
[31;1mError: [0m[1myou've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues[0m
```

If anyone wants to see the full output it is [here](https://pastebin.com/CRnYYdY2) (pretty) or [here](https://pastebin.com/raw/CRnYYdY2) (raw)

# Technical stats

I have seen that someone else had the same error, but it has been sitting for a while and I am on a different OS.

CODE:
```ruby
require "http/server"

server = HTTP::Server.new do |context|
  context.response.content_type = "text/plain"
  context.response.print "Hello world! The time is #{Time.local}"
end

address = server.bind_tcp 8080
puts "Listening on http://#{address}"
server.listen
```

I installed the Crystal compiler via brew. Running `crystal -v` returns:
```console
Crystal 0.35.1 (2020-06-19)

LLVM: 10.0.0
Default target: x86_64-apple-macosx
```

I am on a MacBook Pro 13-inch Early 2011 running macOS 1.14 Mojave.

**Comment by bcardiff** [comment]
Could it be the case that you were running two instances of the compiler at the same time?

**Comment by gaetgu** [comment]
No. I wasn't however, the issue seems to randomly appear and disappear. I have made sure that I was only running one terminal, one tab, one compiler, etc., but it seems to be completely random.

**Comment by brunto** [comment]
Same problem for me. It's been happening since I've been on Big Sur.

**Comment by robacarp** [comment]
I ran into this bug randomly, after having made no change to my system (still running Catalina).

Clearing the ~/.cache/crystal directory provided no benefit, and neither did wiping and reinstalling the shards `lib` directory.

After the comment above regarding multiple instances of the compiler at the same time, I made this change to my Procfile and the error has gone away:

```diff
system_check: script/system_check && sleep 100000
web: lucky watch --reload-browser
assets: yarn watch
-worker: crystal src/mosquito.cr
+worker: sleep 10 ; echo "Starting worker" ; crystal src/mosquito.cr
```

The entrypoint src/mosquito.cr is _barely_ different from the entrypoint start_server.cr which lucky uses. I'm sure the landslide majority of the code pulled from shards lib dir exists in both. In my case a teeplate file was repeatedly causing this error. It would happen first from the web compiler, and then from the worker, etc.

The worker doesn't start compiling _necessarily_ after the lucky version completes, but apparently it's a long enough delay that the race condition is bypassed.

**Comment by InstanceOfMichael** [comment]
This randomly started happening to me today on `Ubuntu 18.04.5 LTS`. Turns out I was running the same specs in 2 different terminals by accident.

**Comment by docelic** [comment]
Solved by #7868 ?

**Comment by asterite** [comment]
Fixed by #10050
